### PR TITLE
Experimental Petset Implementation

### PR DESCRIPTION
**Pull request by olynch**
I have copied some of the features of Petri.jl into a format that uses CSets. I am now thinking that we could move this into AlgebraicPetri. I would like some feedback first, however, and I would like to be more clear on the goals of this whole endeavor.

### CODE REVIEW

**Comment by jpfairbanks** [comment]
Yeah this looks great, I agree that it should be merged into AlgPetri instead of Catlab

**Comment by jpfairbanks** [comment]
We could also extend this definition of a petri net with the initial value problem as a reaction network. The C-Set

```julia
@present ReactionNet(FreeCategory) begin
  species::Ob
  inputs::Ob
  transitions::Ob
  outputs::Ob
  number::Ob

  is::Hom(inputs,species)
  it::Hom(inputs,transitions)
  os::Hom(outputs,species)
  ot::Hom(outputs,transitions)
  rate::Hom(transitions, number)
  u0::Hom(species, number)
end
```

Then models of this theory (C-Sets) are initial value problems for this system. And the open C-Sets construction of glueing along species would allow you to keep track of the rates and initial conditions when you do the glueing.

**Comment by olynch** [comment]
Could we do that with theory extension?

**Comment by epatters** [comment]
You could even do

```julia
@present TheoryReactionNet <: TheoryPetriNet begin
  number::Ob

  rate::Hom(transitions, number)
  u0::Hom(species, number)
end
```

and then write generic code that works for both Petri nets and reaction nets on their common interface.

**Comment by epatters** [comment]
Looks like we had the same thought @olynch :)

**Comment by olynch** [comment]
OK, sweet, I'll do a fresh implementation in the AlgebraicPetri repo, along with the initial value problem.

**Comment by olynch** [comment]
I'm not totally sure how `data` works in a CSet. Do you provide the type of the data when you define the *type*, or do you provide the type of the data when you want to create a *value*?

**Comment by jpfairbanks** [comment]
For documentation I think we need both a blog post style for users and a technical write-up intended for CT people. I also don't know how the data field works, trying to figure it out led me to file #199!

**Comment by epatters** [comment]
> I'm not totally sure how `data` works in a CSet. Do you provide the type of the data when you define the _type_, or do you provide the type of the data when you want to create a _value_?

Currently, you assign the data type when create the `CSet` instance, not when you create the CSet type via `CSetType`. See this [example](https://github.com/AlgebraicJulia/Catlab.jl/blob/d1c561a757e87a0425c5d88fec13a3e203b9dcb1/test/categorical_algebra/CSets.jl#L64) from the unit tests.

Arguably it should be the other way around, although this would require packing yet more information into the `CSet` type parameters. Any thoughts on this are welcome.

> On a related note, I don't know what your plans for doing long-form documentation of CSets, but I would be happy to help, with either a blog-post style introduction, or more technical documentation.

You would be my hero if you wrote some decent docs :) Like James said, ultimately we need both forms of docs but if I had to pick one, I'd start with an introductory, blog-style piece. A lack of such expository content is currently a big barrier for potential Catlab users.

File: experiments/PetSet/PetSet.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+PetSet = CSetType(SITOS, index=[:is,:it,:os,:ot])
+
+nspecies(ps::PetSet) = nparts(ps,:species)
+ntransitions(ps::PetSet) = nparts(ps,:transitions)
+ninputs(ps::PetSet) = nparts(ps,:inputs)
+noutputs(ps::PetSet) = nparts(ps,:outputs)
+
+# Need a better name for this
+function from_std_petri(S,T)
```
Comment (mehalter) [inline experiments/PetSet/PetSet.jl:33]: We could also probably have another method call along the lines of `from_std_petri(m::Petri.Model) = from_std_petri(m.S, m.Δ)`

File: experiments/PetSet/PetSet.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            if s in keys(outs)
+                add_parts!(ps,:outputs,outs[s],(os=j,ot=i))
+            end
+        end
+    end
+    ps
+end
+
+function transition_matrix(ps)
+    T = zeros(Int64,(ntransitions(ps),nspecies(ps),2))
```
Comment (mehalter) [inline experiments/PetSet/PetSet.jl:57]: Possibly change this to using `SparseArrays` and `spzeros`

**Comment by epatters** [comment]
Closing this in lieu of https://github.com/AlgebraicJulia/AlgebraicPetri.jl/pull/12. This will be great to have.

**Pull request closed without merging**